### PR TITLE
[CORE]  New compilation options

### PR DESCRIPTION
**Pull request by roigcarlo**
**📝 Description**

This adds some new options for Kratos compilation and changes how the CI performs the compilation. Some of these set of changes should ease the process to convert Kratos and apps to discoverable CMake "modules" (@mpentek).

This also allows to fix the disk problems in the CI/FullDebug

List of new compilation options:

- `EXCLUDE_KRATOS_CORE=[ON/OFF]`: If set to `ON` allows the core to be excluded from a compilation. Default `OFF`
- `EXCLUDE_AUTOMATIC_DEPENDENCIES=[ON/OFF]`: If set to `ON` allows cmake to ignore the application dependencies specified in the cmake file of the applications. Default `OFF` 

List of changes made to the cmake files:

- Kratos core can now be included as a pre-compiled library which will be searched in the default install directory.
- Trilinos can now be find even if trilinos application is not selected
- `MappingMPIExtension` was renamed to `MeshTrilinosExtension`, as was mainly as it relies on the `EPETRA_EV` vector. Note that with this change CoSimApplication is the only a application with a pure non trilinos mpi extension
- Added explicit include trilinos dir to the extension projects
- Rom Application now correctly adds includes from Eigen/MKL/LinearSolvers

List of changes in the CI:

- Compilation has been layered into stages. At the end of each stage, C++ artifacts are cleaned. With the new compilation options is possible to maintain the libraries 
    - First stage compiles the core and direct core components (Linear Solvers, Metis and Trilinos)
    - Second stage compiles utilities ( No physics apps like: Mapping, Meshing, HDF5, MED, etc... ) 
    - Third stage compiles core applications ( List is not final )
    - Fourth stage compiles the rest.

### CODE REVIEW

**Comment by loumalouomega** [comment]
> **📝 Description** Live testing for the new compilation options for linux
> 
> @loumalouomega Do not approve or merge this, its only for testing against the real CI.

It is not fair :(

**Comment by roigcarlo** [comment]
Close as I finally managed to test the CI locally :D

**Comment by roigcarlo** [comment]
@philbucher This PR may have implications about your recent PR with the selective compilation

**Comment by roigcarlo** [comment]
Na, just making sure everything works, when we have the final list I'll put it here.

I would say that once we decide the apps we will merge, if I manage to solve the issues before, it can be merged and the move the apps accordingly latter.

TBH, I wanted to have it working today, but mpi problems ....

**Comment by philbucher** [comment]
> Na, just making sure everything works, when we have the final list I'll put it here.
> 
> I would say that once we decide the apps we will merge, if I manage to solve the issues before, it can be merged and the move the apps accordingly latter.

Then probably better if I wait. Feel free to review it though ;) 
#11925

> 
> TBH, I wanted to have it working today, but mpi problems ....

Is trilinos not compiling? Check [this commit](https://github.com/KratosMultiphysics/Kratos/pull/11925/commits/2518c2969473778af0c55c4fbb26399a3e05be53) which fixed it for me, there is some manual parsing (which is IMO not very generic/robust) going on which can mess things up:
https://github.com/KratosMultiphysics/Kratos/blob/dcf62e1a5d4dd6bc8f96988b1f7c36b317ab982d/cmake_modules/FindTRILINOS.cmake#L16

**Comment by roigcarlo** [comment]
Nono, is not trilinos, now is PARMMG with gcc

File: applications/MappingApplication/tests/blade_mapping_test.py
Comment (philbucher) [inline applications/MappingApplication/tests/blade_mapping_test.py]: nooo, please dont change this :D

I know it technically still depends on trilinos, but the plan is to replace with the Kratos-CSR matrix to drop the dependency entirely!

File: applications/MeshMovingApplication/trilinos_extension/CMakeLists.txt
Comment (philbucher) [inline applications/MeshMovingApplication/trilinos_extension/CMakeLists.txt:18]: Are you sure this is necessary? The `KratosTrilinosCore` target already does this, IMO it should propagate. Since it worked so far I think the problem is somewhere else

File: applications/RomApplication/CMakeLists.txt
Comment (philbucher) [inline applications/RomApplication/CMakeLists.txt:1]: basically same comment, those changes dont seem right. This should also work without, as in the past

File: applications/MappingApplication/tests/blade_mapping_test.py
Comment (roigcarlo) [reply applications/MappingApplication/tests/blade_mapping_test.py]: Ok... but then we should change the others to MPI extension instead of Trilinos, otherwise is super confusing

**Comment by ddiezrod** [comment]
pinging @aaunnam (@altair devops) just for  you to be aware this, We might need to check this works in  our CI

File: applications/MeshMovingApplication/trilinos_extension/CMakeLists.txt
Comment (roigcarlo) [reply applications/MeshMovingApplication/trilinos_extension/CMakeLists.txt:18]: Yep it is necessary,

As fas as I understand, when CMake resolves the include and linking directories, it takes into account the list of include and linking dirs added with `target_include_directories` from all the projects it has configured during an execution.

Before this changes, we were compiling trilinos along with all other major apps, hence, when including KratosTrillinosCore as a target dependency of, for instance, KratosFluidDynamicsTrilinosExtension, the include was propagated.

After the changes, the compilation of applications can be isolated in different runs, provided that the library exists. This would allow you do to something like:

```
-DKRATOS_ROOT="/already/compiled/kratos"
-DKRATOS_FLUID_DYNAMICS_ROOT="/already/compiled/fluidDynamics"
etc...
```

Which is similar to what is happening with the tiering (the only difference is that the linking directory is currently hardcoded to the install dir, but that could change in the future). With this situation the problem comes the moment the library is added as a linking dependency without having information about how it was created. Then the `target_include_directories` was never executed and cmake has no idea on where to find the headers from trilinos (which are used mainly in the python interface)

File: applications/MappingApplication/tests/blade_mapping_test.py
Comment (roigcarlo) [reply applications/MappingApplication/tests/blade_mapping_test.py]: Reverted the changes :+1:

File: applications/MeshMovingApplication/trilinos_extension/CMakeLists.txt
Comment (roigcarlo) [reply applications/MeshMovingApplication/trilinos_extension/CMakeLists.txt:18]: To be honest, now that I wrote the response I realize that mostly the TrilinosExtensions are dependent on Trilinos includes alone, not TrilinosApplication. Probably linking the whole application into the extension is not the optimal solution....

**Comment by loumalouomega** [comment]
> pinging @aaunnam (@KratosMultiphysics/altair devops) just for you to be aware this, We might need to check this works in our CI

I  modified your comment because you were mentioning a random guy which username is altair

**Comment by roigcarlo** [comment]
ouch sorry

**Comment by roigcarlo** [comment]
@philbucher Its alive! finally.

@ddiezrod If you give me green light from altair and there is no other problem we can merge :+1:

**Comment by ddiezrod** [comment]
@aaunnam Can you please check if our CI passes with these changes?

**Comment by philbucher** [comment]
cool!
I will take a close look, I would appreciate if you can hold of with merging

File: .github/workflows/ci-dummy.yml
Comment (philbucher) [inline .github/workflows/ci-dummy.yml:1]: Why did you delete this one?
I think conceptually it is still needed

Or should we do this similarly to the selective compilation that I am working on?

File: .github/workflows/ci.yml
Comment (philbucher) [inline .github/workflows/ci.yml:327]: can you please also remove it from the docker image right away? We are running low on space with it all the time anyway

**Comment by roigcarlo** [comment]
> cool! I will take a close look, I would appreciate if you can hold of with merging

sure np

**Comment by roigcarlo** [comment]
@philbucher, @@aaunnam did you have time to look over the changes?

**Comment by loumalouomega** [comment]
> @philbucher, @aaunnam did you have time to look over the changes?

A priori does not break our CI

File: .github/workflows/ci.yml
Comment (philbucher) [inline .github/workflows/ci.yml:47]: so far we used 2, can you briefly explain why we can use 4 now?

**Review by philbucher — COMMENTED** [review]
only preliminary comments, still need a bit more time

File: .github/workflows/ci.yml
Comment (roigcarlo) [reply .github/workflows/ci.yml:47]: It is what they offer now:

https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

File: .github/workflows/ci-dummy.yml
Comment (roigcarlo) [reply .github/workflows/ci-dummy.yml:1]: It no longer works as we used it, on + pull_request now marks all files as changed:
https://github.com/actions/runner/issues/2324

File: .github/workflows/ci.yml
Comment (roigcarlo) [reply .github/workflows/ci.yml:327]: Done

File: .github/workflows/configure_core.sh
Comment (philbucher) [inline .github/workflows/configure_core.sh:25]: shouldnt this be in dependencies? 
`configure_core.sh` is a bit misleading otherwise

File: .github/workflows/ci.yml
Comment (philbucher) [inline .github/workflows/ci.yml]: `DUSE_EIGEN_MKL=ON -DUSE_EIGEN_FEAST=ON` are required for the `LinearSolversApp`. As is, it wont build pardiso!
Please move it to the corresponding step (see comment above)

**Review by philbucher — COMMENTED** [review]
nice! Great Job
Only one small correction required

I dont like the code duplication that is required (gitlab ci is soooo much nicer for this) but thats not for this PR. Unless you have an idea/solution for it

File: .github/workflows/configure_core.sh
Comment (roigcarlo) [reply .github/workflows/configure_core.sh:25]: Yes.. I agree, but for reasons I need those three to be compiled in the first stage no matter what, otherwise the list of dependencies and when to activate them becomes a mess. I may change the name of the script to relfect that :S

(bascially moving trilinos and linear solvers retriggers the compilation of the core no matter what, so ....)

File: .github/workflows/ci.yml
Comment (roigcarlo) [reply .github/workflows/ci.yml]: Ouch, good catch

**Review by philbucher — APPROVED** [review]
Lets gooo
Clap 👏🏻 

Btw is this mainly to solve the space issue?
Is it faster or slower now?

And will you also do this for the other apps?

**Comment by roigcarlo** [comment]
It solves the space issue, no change in speed I am afraid, but since we use 4 cores by default maybe its faster.

As for other apps, we can split further if we run out of space again, we can even create a single job for every app and compile selectively is that's what you are asking :)

**Pull request merged**
